File: python/cleaner.py
"""
POCleaner — versi web, menggunakan pipeline_pg (PostgreSQL cache).
"""

import openpyxl
from datetime import datetime


class POCleaner:
    def __init__(self):
        from pipeline_pg import CleansingPipeline, normalize
        self._normalize = normalize
        self.raw_data = []
        self.clusters = []
        self.decisions = {}
        self.filepath = None
        self.header_row = []
        self.desc_col_index = None
        self.desc_col_name = None
        self.available_columns = []
        self._pipeline = CleansingPipeline()
        self._result_map = {}
        self._pipeline_stats = {}

    def load_excel(self, filepath):
        try:
            wb = openpyxl.load_workbook(filepath, data_only=True)
            ws = wb.active
            headers = [str(c.value).strip() if c.value else ""
                       for c in next(ws.iter_rows(min_row=1, max_row=1))]
            self.header_row = headers
            self.available_columns = [h for h in headers if h]
            self.raw_data = []
            for row in ws.iter_rows(min_row=2, values_only=True):
                if all(v is None for v in row): continue
                row_dict = {headers[i]: row[i] for i in range(min(len(headers), len(row)))}
                row_dict["_row_id"] = len(self.raw_data)
                self.raw_data.append(row_dict)
            self.filepath = filepath
            self.clusters = []; self.decisions = {}; self._result_map = {}
            self.desc_col_name = None; self.desc_col_index = None
            auto = None
            kw = ["deskripsi","description","material","nama","uraian","item","barang"]
            for h in self.available_columns:
                if any(k in h.lower() for k in kw): auto = h; break
            return {"ok": True, "message": f"{len(self.raw_data)} baris dimuat.",
                    "count": len(self.raw_data), "columns": self.available_columns,
                    "auto_detect": auto}
        except Exception as e:
            return {"ok": False, "message": str(e)}

    def set_desc_column(self, col_name):
        if col_name not in self.header_row:
            return {"ok": False, "message": f"Kolom '{col_name}' tidak ditemukan."}
        self.desc_col_name = col_name
        self.desc_col_index = self.header_row.index(col_name)
        for row in self.raw_data:
            val = row.get(col_name, "")
            row["_normalized"] = self._normalize(str(val) if val else "")
        self.clusters = []; self.decisions = {}; self._result_map = {}
        return {"ok": True}

    def run_pipeline(self, api_config, threshold=75, progress_cb=None, skip_llm=False):
        if not self.desc_col_name: return [], 0
        self._result_map = self._pipeline.run(
            self.raw_data, self.desc_col_name, api_config,
            progress_cb=progress_cb, skip_llm=skip_llm)
        self._pipeline_stats = self._pipeline.last_stats()
        self.clusters = self._convert_clusters(self._pipeline.last_clusters())
        return self.clusters, sum(1 for c in self.clusters if c["is_duplicate"])

    def _convert_clusters(self, raw_clusters):
        from collections import defaultdict
        desc_col = self.desc_col_name
        desc_to_rows = defaultdict(list)
        for r in self.raw_data:
            desc_to_rows[str(r.get(desc_col,"") or "")].append(r)

        result = []
        for cl in raw_clusters:
            members = []
            for orig in cl["orig_descs"]:
                members.extend(desc_to_rows.get(orig, []))

            canonical = cl["suggested_name"]
            method = "FUZZY_AUTO"; confidence = "high"
            reasoning = ""; warnings = []; is_same = True

            for nk in cl["norm_keys"]:
                rm = self._result_map.get(nk)
                if rm:
                    if rm.get("cleaned_text"): canonical = rm["cleaned_text"]
                    method     = rm.get("method", method)
                    confidence = rm.get("confidence", confidence)
                    reasoning  = rm.get("reasoning", "")
                    warnings   = rm.get("warnings", [])
                    is_same    = rm.get("is_same_item", True)
                    break

            ai_result = None
            if method in ("LLM", "CACHE") and canonical:
                ai_result = {"is_same_item": is_same, "confidence": confidence,
                             "canonical_name": canonical, "reasoning": reasoning,
                             "warnings": warnings}

            result.append({
                "id": hash(cl["cluster_id"]) % 100000,
                "cluster_id": cl["cluster_id"],
                "normalized_key": cl["norm_keys"][0] if cl["norm_keys"] else "",
                "items": members,
                "orig_descs": list(set(cl["orig_descs"])),
                "suggested_name": canonical,
                "avg_score": round(cl.get("min_score", 100), 1),
                "is_duplicate": cl["is_duplicate"],
                "is_auto": cl.get("is_auto", False),
                "method": method, "confidence": confidence,
                "ai_result": ai_result,
            })
        result.sort(key=lambda c: (not c["is_duplicate"], c["avg_score"]))
        return result

    def save_decision(self, cluster_id, action, canonical_name):
        self.decisions[cluster_id] = {
            "action": action, "canonical_name": canonical_name,
            "timestamp": datetime.now().isoformat()}

    def get_decision(self, cluster_id):
        return self.decisions.get(cluster_id)

    def apply_decisions(self):
        from pipeline_pg import normalize
        desc_col  = self.desc_col_name or "Deskripsi Material"
        clean_col = f"{desc_col} (Clean)"
        row_final = {}; audit_entries = []

        for cluster in self.clusters:
            decision = self.decisions.get(cluster["id"])
            cid = cluster.get("cluster_id", str(cluster["id"]))
            for item in cluster["items"]:
                rid  = item["_row_id"]
                orig = str(item.get(desc_col,"") or "")
                nk   = item.get("_normalized", normalize(orig))
                pr   = self._result_map.get(nk, {})
                pm   = pr.get("method","UNPROCESSED")
                pc   = pr.get("confidence","")
                if decision and decision["action"] == "merge":
                    final = decision["canonical_name"]; m = "MANUAL_MERGE"
                    al = "MERGED" if orig != final else "MERGED (no change)"
                elif cluster.get("is_auto") and not decision:
                    final = cluster["suggested_name"]; m = pm; al = "AUTO_MERGED"
                else:
                    final = orig; m = pm or "KEPT"
                    al = ("KEPT" if (decision and decision["action"]=="keep_all")
                          else ("AUTO_MERGED" if cluster.get("is_auto") else "UNREVIEWED"))
                row_final[rid] = (final, m, cid, pc)
                audit_entries.append({
                    "row_id": rid+2, "no_po": item.get("No PO",""),
                    "sumber_ru": item.get("Sumber RU",""),
                    "original": orig, "final": final, "action": al,
                    "method": m, "cluster_id": cid, "confidence": pc,
                    "similarity_score": cluster["avg_score"],
                    "timestamp": (decision or {}).get("timestamp",""),
                })

        final_rows = []
        for r in self.raw_data:
            rc = dict(r); rid = r["_row_id"]
            if rid in row_final:
                ft, m, cid, c = row_final[rid]
            else:
                ft = str(r.get(desc_col,"") or ""); m = "UNPROCESSED"; cid = ""; c = ""
            rc[clean_col] = ft; rc["_method"] = m
            rc["_cluster_id"] = cid; rc["_confidence"] = c
            final_rows.append(rc)
        return final_rows, audit_entries

    def stats(self):
        ps = self._pipeline_stats
        return {
            "total_rows": len(self.raw_data),
            "total_clusters": len(self.clusters),
            "dup_clusters": sum(1 for c in self.clusters if c["is_duplicate"]),
            "desc_col": self.desc_col_name,
            "cache_hits": ps.get("cache_hits",0),
            "auto_merged": ps.get("auto_merged",0),
            "llm_clusters": ps.get("llm_clusters",0),
            "llm_batches": ps.get("llm_batches",0),
        }

    def cache_stats(self):
        return self._pipeline.cache.stats()

File: cleaner.js
'use strict'
const { CleansingPipeline, normalize } = require('./pipeline')

class POCleaner {
  constructor() {
    this.rawData        = []
    this.clusters       = []
    this.decisions      = {}
    this.headerRow      = []
    this.descColName    = null
    this.availableCols  = []
    this._pipeline      = new CleansingPipeline()
    this._resultMap     = {}
    this._pipelineStats = {}
  }

  loadExcel(workbook) {
    try {
      const XLSX    = require('xlsx')
      const ws      = workbook.Sheets[workbook.SheetNames[0]]
      const rows    = XLSX.utils.sheet_to_json(ws, { defval: null, raw: false })
      if (!rows.length) return { ok: false, message: 'File kosong.' }

      this.headerRow     = Object.keys(rows[0])
      this.availableCols = this.headerRow.filter(h => h && !h.startsWith('_'))
      this.rawData       = rows.map((r, i) => ({ ...r, _row_id: i }))
      this.clusters      = []; this.decisions = {}; this._resultMap = {}
      this.descColName   = null

      const kw    = ['deskripsi','description','material','nama','uraian','item','barang']
      const auto  = this.availableCols.find(h => kw.some(k => h.toLowerCase().includes(k))) || null
      return { ok: true, count: this.rawData.length,
               columns: this.availableCols, auto_detect: auto }
    } catch(e) { return { ok: false, message: e.message } }
  }

  setDescColumn(col) {
    if (!this.headerRow.includes(col))
      return { ok: false, message: `Kolom '${col}' tidak ditemukan.` }
    this.descColName = col
    for (const r of this.rawData)
      r._normalized = normalize(String(r[col] || ''))
    this.clusters = []; this.decisions = {}; this._resultMap = {}
    return { ok: true }
  }

  async runPipeline(apiConfig, threshold = 75, progressCb, skipLlm = false) {
    if (!this.descColName) return { clusters: [], dupCount: 0 }
    this._resultMap     = await this._pipeline.run(
      this.rawData, this.descColName, apiConfig, progressCb, skipLlm)
    this._pipelineStats = this._pipeline.lastStats()
    this.clusters       = this._convertClusters(this._pipeline.lastClusters())
    const dupCount      = this.clusters.filter(c => c.is_duplicate).length
    return { clusters: this.clusters, dupCount }
  }

  _convertClusters(rawClusters) {
    const descCol    = this.descColName
    const descToRows = {}
    for (const r of this.rawData) {
      const k = String(r[descCol] || '')
      if (!descToRows[k]) descToRows[k] = []
      descToRows[k].push(r)
    }

    const result = []
    for (const cl of rawClusters) {
      const members = cl.orig_descs.flatMap(d => descToRows[d] || [])
      let canonical  = cl.suggested_name
      let method     = 'FUZZY_AUTO', confidence = 'high'
      let reasoning  = '', warnings = [], isSame = true

      for (const nk of cl.norm_keys) {
        const rm = this._resultMap[nk]
        if (rm) {
          if (rm.cleaned_text) canonical = rm.cleaned_text
          method     = rm.method     || method
          confidence = rm.confidence || confidence
          reasoning  = rm.reasoning  || ''
          warnings   = rm.warnings   || []
          isSame     = rm.is_same_item !== false
          break
        }
      }

      const aiResult = ['LLM','CACHE'].includes(method) && canonical
        ? { is_same_item: isSame, confidence, canonical_name: canonical,
            reasoning, warnings }
        : null

      const id = Math.abs(hashStr(cl.cluster_id)) % 100000
      result.push({
        id, cluster_id: cl.cluster_id,
        norm_keys: cl.norm_keys,
        orig_descs: [...new Set(cl.orig_descs)],
        suggested_name: canonical,
        avg_score: Math.round(cl.min_score * 10) / 10,
        is_duplicate: cl.is_duplicate,
        is_auto: cl.is_auto || false,
        method, confidence, ai_result: aiResult,
        items: members,
        items_count: members.length,
        ru_list: [...new Set(members.map(r => String(r['Sumber RU'] || '')).filter(Boolean))].sort()
      })
    }
    result.sort((a, b) => (b.is_duplicate - a.is_duplicate) || (a.avg_score - b.avg_score))
    return result
  }

  saveDecision(clusterId, action, canonicalName) {
    this.decisions[clusterId] = { action, canonical_name: canonicalName,
                                   timestamp: new Date().toISOString() }
  }

  getDecision(clusterId) { return this.decisions[clusterId] || null }

  applyDecisions() {
    const descCol  = this.descColName || 'Deskripsi Material'
    const cleanCol = `${descCol} (Clean)`
    const rowFinal = {}
    const audit    = []

    for (const cluster of this.clusters) {
      const dec = this.decisions[cluster.id]
      const cid = cluster.cluster_id || String(cluster.id)

      for (const item of cluster.items) {
        const rid  = item._row_id
        const orig = String(item[descCol] || '')
        const nk   = item._normalized || normalize(orig)
        const pr   = this._resultMap[nk] || {}
        const pm   = pr.method || 'UNPROCESSED'
        const pc   = pr.confidence || ''

        let final, method, actionLabel
        if (dec?.action === 'merge') {
          final = dec.canonical_name; method = 'MANUAL_MERGE'
          actionLabel = orig !== final ? 'MERGED' : 'MERGED (no change)'
        } else if (cluster.is_auto && !dec) {
          final = cluster.suggested_name; method = pm; actionLabel = 'AUTO_MERGED'
        } else {
          final = orig; method = pm || 'KEPT'
          actionLabel = dec?.action === 'keep_all' ? 'KEPT'
                      : cluster.is_auto ? 'AUTO_MERGED' : 'UNREVIEWED'
        }

        rowFinal[rid] = { final, method, cid, conf: pc }
        audit.push({ row_id: rid + 2, no_po: item['No PO'] || '',
                     sumber_ru: item['Sumber RU'] || '', original: orig,
                     final, action: actionLabel, method, cluster_id: cid,
                     confidence: pc, similarity_score: cluster.avg_score,
                     timestamp: dec?.timestamp || '' })
      }
    }

    const finalRows = this.rawData.map(r => {
      const row  = { ...r }
      const info = rowFinal[r._row_id]
      row[cleanCol]       = info ? info.final : String(r[descCol] || '')
      row._method         = info?.method || 'UNPROCESSED'
      row._cluster_id     = info?.cid || ''
      row._confidence     = info?.conf || ''
      return row
    })
    return { finalRows, audit }
  }

  stats() {
    const ps = this._pipelineStats
    return {
      total_rows:     this.rawData.length,
      total_clusters: this.clusters.length,
      dup_clusters:   this.clusters.filter(c => c.is_duplicate).length,
      desc_col:       this.descColName,
      cache_hits:     ps.cache_hits     || 0,
      auto_merged:    ps.auto_merged    || 0,
      llm_clusters:   ps.llm_clusters   || 0,
      llm_batches:    ps.llm_batches    || 0,
    }
  }

  async cacheStats() { return this._pipeline.cache.stats() }
}

function hashStr(s) {
  let h = 0
  for (let i = 0; i < s.length; i++) h = Math.imul(31, h) + s.charCodeAt(i) | 0
  return h
}

module.exports = { POCleaner }
